lib/Dialect/SV/Transforms/PrettifyVerilog.cpp

```diff
@@ -52,6 +55,48 @@ static bool isVerilogUnaryOperator(Operation *op) {
   return false;
 }
 
+/// Sink an operation into the same block where it is used.  This will clone the
+/// operation so it can be sunk into multiple blocks. If there are no more uses
+/// in the current block, the op will be removed.
+void PrettifyVerilogPass::sinkOpToUses(Operation *op) {
+  assert(mlir::MemoryEffectOpInterface::hasNoEffect(op) &&
+         "Op with side effects cannot be sunk to its uses.");
+  auto block = op->getBlock();
+  for (auto it = op->use_begin(), end = op->use_end(); it != end;) {
+    // If the current use is not in the same block as the operation, we are
+    // going to clone the op into the use's block.
+    auto &use = *it;
+    auto useBlock = use.getOwner()->getBlock();
+    if (block == useBlock) {
+      ++it;
+      continue;
+    }
+    // Clone the operation and insert it to the beginning of the block.
+    auto *cloned = OpBuilder::atBlockBegin(useBlock).clone(*op);
+    // Scan the use-list for any more uses in the local block and modify them to
+    // use the newly cloned operation. This will only scan forward from the
+    // current use list iterator.
+    for (auto scanIt = std::next(it); scanIt != end;) {
```

Comment:
This is N^2 in the size of teh use list when there are no duplicate blocks.  We don't have a lot of blocks, but i'd recommend just keeping track of previously handled blocks, something like:
`SmallDenseMap<Block, Value> alreadyHandledBlocks;`

outside the loop.  Each time you get to a case where you need to user in another block, just check to see if you've already done it and use the presunk value.  This will make it O(n)

Reply:
I wasn't sure which way to go with this, `n^2` or create a `set`.  The other benefit of a set is I can use `make_early_inc_range` :) (small algo detail was preventing it before).

Reply:
Pushed an update